File: 59按之字形顺序打印二叉树.js
/* function TreeNode(x) {
    this.val = x;
    this.left = null;
    this.right = null;
} */
function Print(pRoot)
{
    // write code here
    if(!pRoot) return [];
    var queue = [], res = [], temp = [];
    var node, num = 0, toBePrinted = 1, isOdd = true;
    queue.push(pRoot);
    while(queue.length) {
        node = queue.shift();
        if(isOdd) {
            temp.push(node.val);
        } else {
            temp.unshift(node.val);
        }
        if(node.left) {
            queue.push(node.left);
            num++;
        }
        if(node.right) {
            queue.push(node.right);
            num++;
        }
        toBePrinted--;
        if(toBePrinted == 0) {
            res.push(temp);
            temp = [];
            toBePrinted = num;
            num = 0;
            isOdd = !isOdd;
        }
    }
    return res;
}
// num记录了同层需要打印的数量 isOdd是当前奇偶的判断
// push向[]末尾添加
// unshift向队首添加一个元素
// shift返回[]第一个元素
// 奇数往队尾加表示从左到右打印 偶数加到队头 表示从右向左打印

File: 59按之字形顺序打印二叉树.java
/*
public class TreeNode {
    int val = 0;
    TreeNode left = null;
    TreeNode right = null;

    public TreeNode(int val) {
        this.val = val;

    }

}
*/
// 从队尾添加元素 从队首取元素 实现先进先出 层次遍历
// 使用迭代器的descendingIterator实现逆序访问
public class Solution {
    public ArrayList<ArrayList<Integer> > Print(TreeNode pRoot) {
        if(pRoot == null) return new ArrayList();
        
        ArrayList<ArrayList<Integer>> res = new ArrayList<>();
        ArrayList<Integer> list = new ArrayList<>();
        LinkedList<TreeNode> queue = new LinkedList<>();
        queue.addLast(null); // regard null as a mark of separate layer
        queue.addLast(pRoot);
        boolean left2right = true;
        
        while(queue.size() != 1) {
            TreeNode node = queue.removeFirst();
            if(node == null) {
                Iterator<TreeNode> iter = null;
                if(left2right) {
                    iter = queue.iterator();
                } else {
                    iter = queue.descendingIterator();
                }
                left2right = !left2right;
                while(iter.hasNext()) {
                    TreeNode temp = (TreeNode)iter.next();
                    list.add(temp.val);
                }
                res.add(new ArrayList<Integer>(list));
                list.clear();
                queue.addLast(null);
                continue;
            }
            if(node.left != null) {
                queue.addLast(node.left);
            }
            if(node.right != null) {
                queue.addLast(node.right);
            }
        }
        return res;
    }
}